### PR TITLE
Small bug fixes

### LINKED ISSUE #3673: `latest-tag-button`, `default-branch-button` and `list-prs-for-file` are not fast/reliable enough
1. Visit https://github.com/sindresorhus/refined-github/blob/42dfa16cd5366de26b475de94f8225003df07364/source/features/list-prs-for-file.tsx (or any file with an open PR) 
2. Refresh the page
3. Wait for each button to appear, if ever

a212

These are 3 separate features but they affect the same area. I think they should be reviewed to understand why they're so slow and why `latest-tag-button` sometimes doesn't appear at all. 

`default-branch-button` is the most annoying one because it's a simple feature with data that's easily available, it should be instant.

### LINKED ISSUE #3945: Copy button doesn't show (and other file-related features)
E.g.: https://github.com/TRSasasusu/SpikingMoYF/blob/master/network.py
image

### PR DESCRIPTION
**Pull request by yakov116**
Fixes #3945

Make features faster closes #3673

- Make `list-prs-for-file` faster
- Make `default-branch-button` faster
- Move addAfterBranchSelector to its own file
- Make `convert-release-to-draft` faster
- Fix `comment-time-machine-links` date check - https://github.com/sindresorhus/refined-github/blob/main/source/features/action-used-by-link.tsx?rgh-link-date=2021-04-28T07%3A12%3A39Z
- Fix duplication in show-whitespace
- Remove outdated comment
- Correct `show-associated-branch-prs-on-fork` display text
- Fix console error in clean-conversation-filters
- Fix types
- Fix console error in faster-pr-diff-options - https://github.com/pixiebrix/pixiebrix-extension/pull/462/files/1e4b6cd476a9bd7f4d5754a86b09087130991fc6..9377c3db676776e997a3317556bd97d5ef62635f
- Sort `show-associated-branch-prs-on-fork` - Will get the last 100
- Exclude issue links in `mark-merge-commits-in-list` - https://github.com/pixiebrix/pixiebrix-extension/commits/main

### CODE REVIEW

File: source/features/latest-tag-button.tsx [hunk 1]
```diff
@@ -125,7 +100,7 @@ async function init(): Promise<false | void> {
 	});
 
 	const link = (
-		<a className="btn btn-sm btn-outline ml-0 flex-self-center css-truncate rgh-latest-tag-button" href={String(url)}>
+		<a className="btn btn-outline ml-0 flex-self-center css-truncate rgh-latest-tag-button" href={String(url)}>
```
Comment (fregante) [inline source/features/latest-tag-button.tsx:103]: 👎 Duplicate of https://github.com/sindresorhus/refined-github/pull/3356

File: source/features/list-prs-for-file.tsx [hunk 2]
```diff
@@ -48,7 +48,7 @@ function getSingleButton(prNumber: number, _?: number, prs?: number[]): HTMLElem
 	return (
 		<a
 			href={getPRUrl(prNumber)}
-			className={'btn btn-sm btn-outline flex-self-center rgh-list-prs-for-file' + (prs ? ' BtnGroup-item' : '')}
+			className={'btn btn btn-outline flex-self-center rgh-list-prs-for-file' + (prs ? ' BtnGroup-item' : '')}
```
Comment (fregante) [inline source/features/list-prs-for-file.tsx:51]: Duplicate of https://github.com/sindresorhus/refined-github/pull/3356

File: source/features/list-prs-for-file.tsx [hunk 3]
```diff
@@ -102,7 +102,7 @@ const getPrsByFile = cache.function(async (): Promise<Record<string, number[]>>
 
 async function init(): Promise<void> {
 	// `clipboard-copy` on blob page, `#blob-edit-path` on edit page
-	const path = select('clipboard-copy, #blob-edit-path')!.getAttribute('value')!;
+	const path = (await elementReady('clipboard-copy, #blob-edit-path', {waitForChildren: false}))!.getAttribute('value')!;
```
Comment (fregante) [inline source/features/list-prs-for-file.tsx:105]: Check the elements it selects: If they don't have much content inside, `waitForChildren: false` doesn't make any difference.

File: source/features/show-whitespace.tsx [hunk 4]
```diff
@@ -76,5 +76,6 @@ void features.add(__filebasename, {
 		onNewComments,
 		onPrFileLoad
 	],
+	deduplicate: '.rgh-observing-whitespace',
```
Comment (fregante) [inline source/features/show-whitespace.tsx:79]: Won't this break `additionalListeners`? 

`.rgh-observing-whitespace` matches any observed line, which may still be on the page when this check runs.

I think we'll need an additional class probably on the top-most element that is replaced (`#pjax-container > .rgh-observing-whitespace:first-child` ? not sure)

**Review by fregante — CHANGES_REQUESTED** [review]

**Comment by fregante** [comment]
#4444 nice.

**Comment by fregante** [comment]
Can you also drop the comment at the top of refined-github.ts since we’re going to use hotfixes from now on?

File: source/features/latest-tag-button.tsx (hunk 1 above)
Comment (yakov116) [reply source/features/latest-tag-button.tsx:103]:
```suggestion
		<a className="btn btn-sm btn-outline ml-0 flex-self-center css-truncate rgh-latest-tag-button" href={String(url)}>
```

File: source/features/list-prs-for-file.tsx (hunk 3 above)
Comment (yakov116) [reply source/features/list-prs-for-file.tsx:105]:
```suggestion
	const path = (await elementReady('clipboard-copy, #blob-edit-path'))!.getAttribute('value')!;
```

File: source/features/list-prs-for-file.tsx (hunk 2 above)
Comment (yakov116) [reply source/features/list-prs-for-file.tsx:51]:
```suggestion
			className={'btn btn-sm btn-outline flex-self-center rgh-list-prs-for-file' + (prs ? ' BtnGroup-item' : '')}
```

File: source/features/highlight-deleted-and-added-files-in-diffs.tsx [hunk 5]
```diff
@@ -40,7 +40,7 @@ async function init(): Promise<void | false> {
 			const sourceIcon = pageDetect.isPR() ?
 				select(`[href="${filename.hash}"] svg`, fileList)! :
 				select(`svg + [href="${filename.hash}"]`, fileList)?.previousElementSibling!;
```
[the remark this answers is not in the document]
Comment (yakov116) [reply source/features/highlight-deleted-and-added-files-in-diffs.tsx:42]: The issue is the cloneNode

File: source/features/show-whitespace.tsx (hunk 4 above)
Comment (yakov116) [reply source/features/show-whitespace.tsx:79]: > Won't this break `additionalListeners`?

I do not think so its not excluded in the deduplication

File: source/features/highlight-deleted-and-added-files-in-diffs.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (fregante) [reply source/features/highlight-deleted-and-added-files-in-diffs.tsx:42]: That's unlikely, you can see here that the cloned node () is of the correct type without assertions


https://github.com/sindresorhus/refined-github/blob/8ca75a97bf92ade57b220f9945a06e3eee865c9c/source/features/conversation-filters.tsx#L20-L26

File: source/features/show-whitespace.tsx (hunk 4 above)
Comment (fregante) [reply source/features/show-whitespace.tsx:79]: I think we had this discussion before. You're right 😃

**Comment by fregante** [comment]
What's https://github.com/sindresorhus/refined-github/commit/b45b4bb3f665337fb7a3dddda50d64693a8aa55e? It's not mentioned in the post and there are no test links or bug issues

**Comment by yakov116** [comment]
> What's [b45b4bb](https://github.com/sindresorhus/refined-github/commit/b45b4bb3f665337fb7a3dddda50d64693a8aa55e)? It's not mentioned in the post and there are no test links or bug issues

I marked it as draft. My computer broke (son dropped it 😢 ), I have to redo the whole first post.

**Comment by yakov116** [comment]
That checks if the date of the comment is after or before the last commit.

The selector needed to be updated.

**Comment by fregante** [comment]
> I marked it as draft

Nope 😃 

Screen Shot


> son dropped it 😢

No worries, my dog ate my homework once too. 😜 Take your time of course

File: source/features/highlight-deleted-and-added-files-in-diffs.tsx [hunk 6]
```diff
@@ -39,7 +39,7 @@ async function init(): Promise<void | false> {
 			filename.classList.add('rgh-pr-file-state');
 			const sourceIcon = pageDetect.isPR() ?
 				select(`[href="${filename.hash}"] svg`, fileList)! :
-				select(`svg + [href="${filename.hash}"]`, fileList)?.previousElementSibling!;
+				select(`svg + [href="${filename.hash}"]`, fileList)?.previousElementSibling as SVGSVGElement;
```
Comment (fregante) [inline source/features/highlight-deleted-and-added-files-in-diffs.tsx:42]: SVG elements don't have `href` attributes, this is not `<SVG>` being selected here. Is the selector wrong?

File: source/features/clean-conversation-filters.tsx [hunk 7]
```diff
@@ -55,7 +55,7 @@ async function hasProjects(): Promise<boolean> {
 
 async function hideProjects(): Promise<void> {
 	if (!await hasProjects()) {
-		(await elementReady('[data-hotkey="p"]'))!.parentElement!.remove();
+		(await elementReady('[data-hotkey="p"]'))?.parentElement!.remove();
```
Comment (fregante) [inline source/features/clean-conversation-filters.tsx]: Line 57 exactly checks for projects. If we get to line 58, there should not be any projects.

Comment (yakov116) [reply source/features/clean-conversation-filters.tsx]: hmm I will double check what's going on

**Review by fregante — APPROVED** [review]
Other than `source/features/comments-time-machine-links` and the reverted commit, the code so far is approved.

File: source/features/clean-conversation-filters.tsx (hunk 7 above)
Comment (yakov116) [reply source/features/clean-conversation-filters.tsx]: > Line 57 exactly checks for projects.

Correct but it does not check if the projects dropdown exists.
I have the error on https://github.com/download-directory/download-directory.github.io/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Abug

**Comment by yakov116** [comment]
> Other than `source/features/comments-time-machine-links`,

example url that should not show the bar
```
https://github.com/sindresorhus/refined-github/blob/main/source/features/action-used-by-link.tsx?rgh-link-date=2021-04-28T07%3A12%3A39Z
```

**Comment by fregante** [comment]
> example url that should not show the bar

I'm confused, why should that page not show the bar?

- It points to a branch
- The URL has a date in April

File: source/features/clean-conversation-filters.tsx (hunk 7 above)
Comment (fregante) [reply source/features/clean-conversation-filters.tsx:58]: Of course it doesn't exist, it doesn't have any projects.

Comment (fregante) [reply source/features/clean-conversation-filters.tsx:58]: That should already be covered:

https://github.com/sindresorhus/refined-github/blob/7c5c2a08129bb2941e1121c457b908e58f3a0d43/source/features/clean-conversation-filters.tsx#L48-L50

Comment (fregante) [reply source/features/clean-conversation-filters.tsx:58]: Let's merge `hideProjects` and `hasProjects` into something that makes more sense. Currently the condition is "has projects" while it tries to be "should it hide the section"

Why are we even checking for projects if there's no sidebar section to hide at all?

File: source/features/mark-merge-commits-in-list.tsx [hunk 8]
```diff
@@ -34,7 +34,7 @@ const filterMergeCommits = async (commits: string[]): Promise<string[]> => {
 
 // eslint-disable-next-line import/prefer-default-export
 export function getCommitHash(commit: HTMLElement): string {
-	return commit.querySelector('a[href]')!.pathname.split('/').pop()!;
+	return commit.querySelector('a[href]:not(.js-issue-link)')!.pathname.split('/').pop()!;
```
Comment (fregante) [inline source/features/mark-merge-commits-in-list.tsx:37]: Does not work? PRs also have the same class

Screen Shot

**Comment by fregante** [comment]
Thanks for the many bug fixes but it's best to open multiple PRs, **following the template.** If I can't see the bug in a link, in a screenshot, or in a issue, I'm reverting.

**Comment by yakov116** [comment]
> Thanks for the many bug fixes but it's best to open multiple PRs, **following the template.** If I can't see the bug in a link, in a screenshot, and/or in a issue, I'm reverting.

Please look at the first post post....

**Comment by fregante** [comment]
Where are the screenshots? I don’t know what “Fix duplication in show-whitespace” refers to. No screenshot of bug, no screenshot of fix, no link to any issues for most fixes here. 

We had discussed that specific bug somewhere but that doesn’t mean everyone reading knows. Follow the template.

File: source/features/mark-merge-commits-in-list.tsx (hunk 8 above)
Comment (yakov116) [reply source/features/mark-merge-commits-in-list.tsx]: PR's? We are catching the wrong link.

From now on I will open an issue for each and thing I want to fix.

Comment (yakov116) [reply source/features/mark-merge-commits-in-list.tsx]: See https://github.com/pixiebrix/pixiebrix-extension/commits/main

Comment (fregante) [reply source/features/mark-merge-commits-in-list.tsx]: > From now on I will open an issue for each and thing I want to fix.

If you're writing a PR, you can include the details in the PR itself to save some time

**Pull request merged**